src/v/storage/disk_log_impl.cc

```diff
@@ -801,6 +807,22 @@ ss::future<compaction_result> disk_log_impl::compact_adjacent_segments(
         }
     }
 
+    bool all_segments_self_compacted = std::all_of(
+      segments.begin(), segments.end(), [](ss::lw_shared_ptr<segment> segment) {
+          return segment->finished_self_compaction();
+      });
+
+    if (!all_segments_self_compacted) {
+        const auto total_size = std::accumulate(
+          segments.begin(),
+          segments.end(),
+          size_t(0),
+          [](size_t acc, ss::lw_shared_ptr<segment>& seg) {
+              return acc + seg->size_bytes();
+          });
+        co_return compaction_result{total_size};
+    }
```

Comment:
looks like the underlying iterator type is not range friendly, so if I move this check to the beginning, I have to undo range::all_of(), retaining it here since this branch is seldom taken and vector size is small.

Reply:
ah right, seastar::circular_buffer. the fix is upstream but we haven't updated yet